app/src/behaviors/behavior_tap_dance.c

```diff
@@ -0,0 +1,340 @@
+static inline int press_tap_dance_behavior(struct active_tap_dance *tap_dance, int64_t timestamp) {
+    // Press chosen tap dance binding
+    struct zmk_behavior_binding binding =
+        tap_dance->config
+            ->behaviors[MIN(tap_dance->counter, tap_dance->config->behavior_count) - 1];
+    struct zmk_behavior_binding_event event = {
+        .position = tap_dance->position,
+        .timestamp = timestamp,
+    };
+    return behavior_keymap_binding_pressed(&binding, event);
+}
+
+// release_tap_dance_behavior
+
+// Returns behavior_keymap_binding_pressed to ensure that keycode is actually released
+static inline int release_tap_dance_behavior(struct active_tap_dance *tap_dance,
+                                             int64_t timestamp) {
+    // Release chosen tap dance binding
+    struct zmk_behavior_binding binding =
+        tap_dance->config
+            ->behaviors[MIN(tap_dance->counter, tap_dance->config->behavior_count) - 1];
+    struct zmk_behavior_binding_event event = {
+        .position = tap_dance->position,
+        .timestamp = timestamp,
+    };
+    clear_tap_dance(tap_dance);
+    return behavior_keymap_binding_released(&binding, event);
+}
+
+// on_tap_dance_binding_pressed
+
+// Code that determines what happens when a tap-dance binding is pressed.
+// Does not register a keycode press unless the tap-dance's counter is
+// equal to the maximum number of bindings associated with it.
+// E.g. if a defined tap-dance has bindings = <&kp N1>, <&kp N2>, <&kp N3>;
+// when the counter is equal to 3, press_tap_dance_behavior will be called
+// to register N3's keycode being pressed
+static int on_tap_dance_binding_pressed(struct zmk_behavior_binding *binding,
+                                        struct zmk_behavior_binding_event event) {
+    // Code executes when the tap dance key is pressed.
+    const struct device *dev = device_get_binding(binding->behavior_dev);
+    const struct behavior_tap_dance_config *cfg = dev->config;
+    struct active_tap_dance *tap_dance;
+    tap_dance = find_tap_dance(event.position);
+    // No active tap dance found on keypress location, attempt to create a new one
+    if (tap_dance == NULL) {
+        if (new_tap_dance(event.position, cfg, &tap_dance) == -ENOMEM) {
+            LOG_ERR("Unable to create new tap dance. Insufficient space in active_tap_dances[].");
+            return ZMK_BEHAVIOR_OPAQUE;
+        }
+        // Tap dance successfully created. Reset timer.
+        LOG_DBG("%d created new tap dance", event.position);
+        reset_timer(tap_dance, event);
+        return ZMK_BEHAVIOR_OPAQUE;
+    }
+    // An active tap dance was found at the keypress location. Stop the timer.
+    tap_dance->is_pressed = true;
+    LOG_DBG("%d tap dance re-pressed", event.position);
+    stop_timer(tap_dance);
+    // Increment the counter. Check if the counter has reached the maximum behavior number.
+    tap_dance->counter++;
```

Comment:
if you check the maximum before incrementing the count, (e.g., move this line inside of the if statement and modify the if statement slightly) you don't run the risk of having a higher counter than there are behavior_counts, removing the need to use `MIN(tap_dance->counter, tap_dance->config->behavior_count) - 1)` everywhere.